File: mo-auth/src/main/java/ink/ckx/mo/auth/util/KeyStoreKeyFactory.java
package ink.ckx.mo.auth.util;

import org.springframework.core.io.Resource;
import org.springframework.util.StringUtils;

import java.io.InputStream;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyStore;
import java.security.PublicKey;
import java.security.cert.Certificate;
import java.security.interfaces.RSAPrivateCrtKey;
import java.security.spec.RSAPublicKeySpec;

/**
 * @author chenkaixin
 * @description
 * @since 2023/10/19
 */
public class KeyStoreKeyFactory {

    private final Resource resource;

    private final char[] password;

    private final Object lock;

    private final String type;

    private KeyStore store;

    public KeyStoreKeyFactory(Resource resource, char[] password) {
        this(resource, password, type(resource));
    }

    public KeyStoreKeyFactory(Resource resource, char[] password, String type) {
        this.lock = new Object();
        this.resource = resource;
        this.password = password;
        this.type = type;
    }

    private static String type(Resource resource) {
        String ext = StringUtils.getFilenameExtension(resource.getFilename());
        return ext == null ? "jks" : ext;
    }

    public KeyPair getKeyPair(String alias) {
        return this.getKeyPair(alias, this.password);
    }

    public KeyPair getKeyPair(String alias, char[] password) {
        try {
            synchronized (this.lock) {
                if (this.store == null) {
                    synchronized (this.lock) {
                        this.store = KeyStore.getInstance(this.type);
                        try (InputStream stream = this.resource.getInputStream()) {
                            this.store.load(stream, this.password);
                        }
                    }
                }
            }

            RSAPrivateCrtKey key = (RSAPrivateCrtKey) this.store.getKey(alias, password);
            Certificate certificate = this.store.getCertificate(alias);
            PublicKey publicKey = null;
            if (certificate != null) {
                publicKey = certificate.getPublicKey();
            } else if (key != null) {
                RSAPublicKeySpec spec = new RSAPublicKeySpec(key.getModulus(), key.getPublicExponent());
                publicKey = KeyFactory.getInstance("RSA").generatePublic(spec);
            }

            return new KeyPair(publicKey, key);
        } catch (Exception var14) {
            throw new IllegalStateException("Cannot load keys from store: " + this.resource, var14);
        }
    }
}

File: mo-auth/src/main/kotlin/ink/ckx/mo/auth/util/KeyStoreKeyFactory.kt
package ink.ckx.mo.auth.util

import org.springframework.core.io.Resource
import org.springframework.util.StringUtils
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyStore
import java.security.PublicKey
import java.security.interfaces.RSAPrivateCrtKey
import java.security.spec.RSAPublicKeySpec

/**
 * @author chenkaixin
 * @description
 * @since 2023/10/19
 */
class KeyStoreKeyFactory @JvmOverloads constructor(
    private val resource: Resource,
    private val password: CharArray,
    private val type: String = type(resource)
) {

    private val lock: Any = Any()
    private var store: KeyStore? = null

    fun getKeyPair(alias: String): KeyPair {
        return this.getKeyPair(alias, password)
    }

    private fun getKeyPair(alias: String, password: CharArray): KeyPair {
        return try {
            synchronized(lock) {
                if (store == null) {
                    synchronized(lock) {
                        store = KeyStore.getInstance(type)
                        resource.inputStream.use { stream -> store!!.load(stream, this.password) }
                    }
                }
            }
            val key = store!!.getKey(alias, password) as RSAPrivateCrtKey
            val certificate = store!!.getCertificate(alias)
            val publicKey: PublicKey?
            if (certificate != null) {
                publicKey = certificate.publicKey
            } else {
                val spec = RSAPublicKeySpec(key.modulus, key.publicExponent)
                publicKey = KeyFactory.getInstance("RSA").generatePublic(spec)
            }
            KeyPair(publicKey, key)
        } catch (var14: Exception) {
            throw IllegalStateException("Cannot load keys from store: $resource", var14)
        }
    }

    companion object {
        private fun type(resource: Resource): String = StringUtils.getFilenameExtension(resource.filename) ?: "jks"
    }
}